### ISSUE TITLE
Decide upon a robust development model and branching structure

### ISSUE THREAD

**Issue by gsrohde** [biocro/biocro#34]
@eloch216:

For long-term general organization, I want to suggest that besides the main branch, we also have an unreleased branch. The unreleased branch can be used for development, and changes can be collected there until we are ready to release a new version. We could have branch protection rules to prevent any changes directly to main; instead, they should only occur via merging from unreleased to main.

(In the long term, I want to remove the module library from biocro/biocro. When that's complete, it's unlikely that any development on biocro/biocro would need to be private, and can simply proceed on the unreleased branch.)

In that case, we might want to have documentation available for the unreleased branch, if only to help developers with making sure it looks correct. Public documentation for the main branch is a higher priority, though.

@gsrohde:

I was just thinking about this [long-term general organization] yesterday. I like the [git-flow](https://nvie.com/posts/a-successful-git-branching-model/) development model, which a colleague from NCSA put me onto 5 years ago when we were working on BETYdb together. (He said they also used it for PEcAn and for other projects at NCSA.)

@justinmcgrath:

I think that is a good workflow. In broad terms, that is what we have been doing with ebimodeling/biocro/main as the "main" branch and ebimodeling/biocro-dev/main as the "develop" branch, but with the large inconvenience of being in separate repositories, and then also a fair bit of chaos with multiple groups working on different "feature" branches that don't get merged regularly.

**Comment by eloch216** [comment]
Another message Justin sent by email:

> I like the workflow in the cite you linked. Our workflow already looks similar. One difference I see from what we have done is that hotfixes are branched off main and merged into develop. I'm not sure how we've been doing that, but I don't think it's been done as cleanly, and it's a nice solution. The workflow also branches from develop into a release, where only bug fixes are allowed. I think a good way to do it. In practice, I don't know that we'll have cases where someone is itching to work on a feature while a release is being tidied up, but if that does happen, I like that solution.
>
> Maybe there are some things to consider regarding coordination with the skeleton library, but offhand I can't think of anything. I also don't see any ways that this would create a burden for anything we've done in the past. Overall, I like the work flow and am happy to copy it entirely.

I also like the `git-flow` development model. Although there seem to be lots of claims that `git-flow` is "outdated," it seems appropriate for us since we are planning to use versioned releases rather than continuous web deployment.

I think there are two steps we can take to get started with this workflow:
1. Create a `develop` branch. (I am happy with this name since it is more common than "unreleased," which I had suggested before I knew about `git-flow`.)
2. Set branch protection rules for `main` and `develop` so they cannot be deleted, and to require pull requests before merging. (In this workflow, I think all commits are supposed to be made on feature, release, or hotfix branches, and then merged into `main` and/or `develop`.)

I will go ahead and do this. It's nice that we can now use branch protection rules since the repository is public.

**Comment by eloch216** [comment]
As a test of my `git-flow` understanding: There will soon be a few changes to the documentation workflows.
1. These don't impact the BioCro software package itself.
2. It's important for them to immediately go into `main`.

So, I think they will be considered a "hotfix." Do you agree?

**Comment by gsrohde** [comment]
> As a test of my `git-flow` understanding: There will soon be a few changes to the documentation workflows.
> 
> 1. These don't impact the BioCro software package itself.
> 2. It's important for them to immediately go into `main`.
> 
> So, I think they will be considered a "hotfix." Do you agree?

Maybe.
> 
> In this case, a new branch would be created based on `main`, and then merged into `main` and `develop` when it is ready. There should be a change in the version number, but only in the third digit. We probably don't need to tag the resulting release. Is this correct?

I'm thinking this is really an exceptional case that isn't addressed by git-flow.  I'm thinking that changes that _only_ affect the workflows and don't affect end users in any way (except where the go to view the published documentation) not only don't need to be tagged, they don't need a new version number either.

**Comment by gsrohde** [comment]
Perhaps we could call the branch `workflow-fix`.

The quick and easy fix will be to simply change the target location of the documentation from ebimodeling to biocro.  To make the latest documentation appear directly at https://biocro.github.io/biocro-documentation instead of at https://biocro.github.io/biocro-documentation/v3.0.0/pkgdown is more involved, and I'll probably use a feature branch to do that work.

**Comment by gsrohde** [comment]
If you want to print out the diagram that goes with the git-flow article and hang it on your wall, it's at https://nvie.com/files/Git-branching-model.pdf.

**Comment by eloch216** [comment]
> I'm thinking this is really an exceptional case that isn't addressed by git-flow. I'm thinking that changes that _only_ affect the workflows and don't affect end users in any way (except where the go to view the published documentation) not only don't need to be tagged, they don't need a new version number either.

That's probably a good way to view it. It's probably a fairly common occurrence, though, at least for us. Maybe we could call it a "webfix" rather than a hotfix, and have separate rules (no version number change, no tag).

I wonder why this type of branch/change isn't included in `git-flow`. Is it somehow possible to have stuff like our documentation workflow running in a separate repository? I'm not suggesting that we do that, I'm just curious.

**Comment by gsrohde** [comment]
> I wonder why this type of branch/change isn't included in `git-flow`. 

Maybe because GitHub workflows weren't around 10 years ago?  Though I would think the same considerations would apply to Travis.

> Is it somehow possible to have stuff like our documentation workflow running in a separate repository? I'm not suggesting that we do that, I'm just curious.

Yes, I think this is possible.  Slightly more complicated though.  And I'm not sure whether you could have builds triggered by repository changes or whether you'd just have to do regular builds on a set schedule.  (I think the GitHub REST API would allow for the former, but I'm not at all familiar with it.)

**Comment by justinmcgrath** [comment]
> If you want to print out the diagram that goes with the git-flow article and hang it on your wall, it's at https://nvie.com/files/Git-branching-model.pdf.

Thanks. We should also include something about this in the contribution guidelines.

> I wonder why this type of branch/change isn't included in git-flow. Is it somehow possible to have stuff like our documentation workflow running in a separate repository? I'm not suggesting that we do that, I'm just curious.

I think it's because the documentation workflow is really a separate, but related project, so it would ideally have its own versions. Putting it in a separate repository would fix that, but I don't know that it's necessary. I would say treat it in the regular manner - so something like this would be a hot fix, and other changes related to features would be in feature branches. I would include bumping version numbers for any workflow change. I can't see what we gain by not doing that, and it certainly seems like an opportunity for confusion.

Here it's a bit frustrating because it would have been nice to do this before a "release" so it didn't seem goofy to immediately have a version change that didn't relate directly to BioCro itself, but I think typically the workflow changes will be happening along with BioCro changes in most cases once we get through the transition to a new organization.

**Comment by justinmcgrath** [comment]
And in hindsight I should have waited a few days to settle into the new organization before tagging a release. I'm sure no one has done anything with this, so we could change the tag after this stuff is working to avoid a version bump.

**Comment by justinmcgrath** [comment]
How do feature branches handle changes on `develop`? The chart implies that feature branches never get changes from develop. Is that is what is intended. You wouldn't get hot fixes, but maybe that's OK depending what you're doing. Potentially it could result in a lot of merge conflicts if a feature is in the works for a long time and other features are started, finished and merged in that time.

**Comment by gsrohde** [comment]
> I would include bumping version numbers for any workflow change. I can't see what we gain by not doing that, and it certainly seems like an opportunity for confusion.

Mainly, what we gain is sparing ourselves the extra work, such as writing up release version notes for a change users could care less about.

In general, though, I would put more extensive, less urgent changes to GitHub workflow files in a feature branch, just like any other change, even if the changes have no effect upon the BioCro software itself.

**Comment by justinmcgrath** [comment]
> In general, though, I would put more extensive, less urgent changes to GitHub workflow files in a feature branch, just like any other change, even if the changes have no effect upon the BioCro software itself.

I agree with that, but if the point of `main` is that it's a series of releases, then any change there should have a new version number. Otherwise, what version is it? If someone cites BioCro using that version, what are they going to write? It's not a huge consideration, but I don't like special cases. They make things complicated - "any time there's a change to `main` we'll change the version number, except when ..., or when ...". It's also not the workflow I think we've all agreed we like: "Therefore, each time when changes are merged back into master, this is a new production release by definition."

Writing a short release note that says it fixed documentation is easy, and the sort of short note you'd expect from a patch version increase.

We obviously don't want to do this sort of thing regularly, and hopefully we'll get to a point where things are organized enough that this happens less often.

**Comment by gsrohde** [comment]
> I agree with that, but if the point of `main` is that it's a series of releases, then any change there should have a new version number. Otherwise, what version is it? If someone cites BioCro using that version, what are they going to write? It's not a huge consideration, but I don't like special cases. 

I was thinking that users should only use tagged releases.  But it is probably simpler just to have a blanket policy that every commit along `main` should have a unique tag and release number.  This way we won't have to split hairs about whether a commit should be tagged or not.  One thing I am adamant about, however, is that every new release should be tagged.

**Comment by gsrohde** [comment]
> And in hindsight I should have waited a few days to settle into the new organization before tagging a release. I'm sure no one has done anything with this, so we could change the tag after this stuff is working to avoid a version bump.

I don't think it's any great loss if the first publicized version ends up being 3.0.1 or 3.0.2.

**Comment by justinmcgrath** [comment]
> If a feature branch needs recent changes from a release branch or a hotfix branch, they in general should first be merged into `develop` and then `develop` should be merged into the feature branch.

Why merge to develop and then merge back as opposed to only merging develop into the feature branch? If you merged into develop first, there'd be an incomplete feature.

**Comment by gsrohde** [comment]
> Why merge to develop and then merge back as opposed to only merging develop into the feature branch? If you merged into develop first, there'd be an incomplete feature in develop, which depending on timing could end up in main.

By "they", I meant release or hotfix branches.  Feature branches should, in general, only be merged into `develop` when they are done.

**Comment by gsrohde** [comment]
> > Although the chart doesn't show it, and as far as I can see, the document doesn't mention it, it seems perfectly fine to me if develop is merged into a feature branch at some point after the feature branch originally branched off of develop. The main case for this that I can see is if some more recent change to develop would affect how the feature is being implemented.
> 
> Since it's not in the page you linked, I think we need to write this in our guidelines.

One thing I'd really like to do regarding documentation is to clearly separate information directed at coders/contributers and that directed at project maintainers.  Right now, that is mostly the same set of people, but I imagine that may not always be the case.

**Comment by justinmcgrath** [comment]
It sounds like we've decided that git flow works well for us. I think the next steps for closing this issue is to write this down somewhere. Maybe in the contribution guidelines could link to the page above for those who want the full reasoning. It may be good to have a checklist of steps. Mostly I'm thinking of hotfixes and releases so we don't forget key things. For hotfixes, it would be something like this.

1. Create a branch from main.
2. Increment the version number in DESCRIPTION.
3. Make necessary changes.
4. Describe the changes in NEWS.md.
5. Create a pull request to merge into `main`.
6. After approval, merge into `main` and `develop`.

For hot fixes, I don't know if we should be creating two separate PRs for each of `main` and `develop`. Potentially the merges won't be the same, so maybe it's good to do, but I'm not sure it's necessary.

An augmentation or possible alternative to check lists in documentation files is putting them in pull request templates. I'm not sure how they work, but maybe you can include a markdown checklist.

**Comment by gsrohde** [comment]
> For hotfixes, it would be something like this.
> 
> 1. Create a branch from main.

This should mention the naming convention for the newly-created branch, e.g. `hotfix-v3.0.2` for a fix that will change the version number to `v3.0.2`.

**Comment by eloch216** [comment]
> > For hotfixes, it would be something like this.
> > 
> > 1. Create a branch from main.
> 
> This should mention the naming convention for the newly-created branch, e.g. `hotfix-v3.0.2` for a fix that will change the version number to `v3.0.2`.

Good idea. I was just going through the branches in this repository to see what I could delete. I found it very helpful when a branch was prefixed with its type (e.g. `feature/`, `hotfix-`, etc). That makes the "completeness" criteria easy to determine -- a hotfix branch should have been merged into `main` and `develop` while a feature branch should only have been merged into `develop`. This naming scheme will also make intentions clear when starting a PR.

**Comment by justinmcgrath** [comment]
If you make a pull request, GitHub says whether there are conflicts or test failures, so creating one and merging without review in the absence of conflicts and test problems seems to be the same amount of effort as doing a local merge and push if there are not conflicts. The PR runs the tests automatically though, and although it seems unlikely that a PR without conflicts would cause a test failure on develop but not main, maybe that's a possibility.

**Comment by gsrohde** [comment]
> If you make a pull request, GitHub says whether there are conflicts or test failures, so creating one and merging without review in the absence of conflicts and test problems seems to be the same amount of effort as doing a local merge and push if there are not conflicts. The PR runs the tests automatically though, and although it seems unlikely that a PR without conflicts would cause a test failure on develop but not main, maybe that's a possibility.

So maybe after merging a hotfix into `main`, we can attempt a merge into `develop` on GitHub, and if there are conflicts (or at least if there are non-trivial conflicts) one could make a PR instead of resolving the conflicts and going ahead with the merge?

**Comment by eloch216** [comment]
If we expect hotfixes to get merged into `develop` without a PR, I will need to change the branch protection rule that I had previously set up; it requires a PR when merging into `develop`. I thought it would be helpful to have this rule, but we can get rid of it if we need to.

In the long run, we probably won't be making hotfixes as often as we are right now. So, personally, I would rather keep the rule and have two PRs for each hotfix. But if others don't think the tradeoff is worth it, I can remove the rule.

**Comment by justinmcgrath** [comment]
I guess I wonder the reason for a strong aversion to PRs. They're easy to make and then complete. On the Web site, it's easier to create a PR than to only merge branches. I'm don't even know how to merge branches on the Web site without a PR. Compared to any of the options, they seem easier to do, and I don't see any downside to them that would offset it. Somewhat tangentially, the different options seem so similar that if someone did any of them, I wouldn't particularly care or notice.

**Comment by gsrohde** [comment]
> I guess I wonder the reason for a strong aversion to PRs. They're easy to make and then complete.

Perhaps I have in mind our history of having PRs languish unattended for weeks on end.  _Maybe_ this wouldn't happen in cases like this, where the code has mostly already been reviewed once, when merged into the `main` branch, but I'm not entirely confident this will be the case.

My interest is in always keeping `develop` up to date with `main`.  One thing I especially want to avoid happening is initiating a new release branch that doesn't have all the latest work from main (which it should get from an up-to-date `develop` branch).

Here's what I had imagined to be the cases that should require PRs:
1. All merges into `main`.
2. Merges from feature branches into `develop`.

I thought it unnecessary to make a PR for merges of release branches or hotfix branches into develop, since the code they contain is already reviewed once for their merging into `main`.

**Comment by justinmcgrath** [comment]
It seems like we agree on all of the big things. What I'm thinking is that you'd make a PR simply because that's a very easy way to do a merge on the Web site. If there are no conflicts or test failures, then it's merged, without review, immediately after merging in `main`, as a matter of course. It's functionally equivalent to a local merge and push without review. I don't see how either of the options leave something unattended. Ed did exactly this earlier today, creating and merging #45 immediately after #44 was approved and merged. For either option, the issue will be remembering to do it.

If there are conflicts, then both a PR from the start and a local merge that is then turned into a PR are equally likely to languish. Then in either case, how long it takes depends on how quickly the conflicts are addressed, and it doesn't matter how it ended up as a PR.

I see almost no difference between the options. Ed did point out the branch protection rules though, which would prevent locally merging and pushing. I like the idea of enforcing PRs for merges to develop, and possibly there's a way to make an exclusion for hotfixes, but it requires less set up to always have a PR to merge to develop.

**Comment by gsrohde** [comment]
Do the branch protection rules require a review and approval?  If so, can this be required for merges from some branches but not others?

I guess I was thinking requiring a PR for merging a hotfix into `develop` necessarily entailed a review, with all the possible delay this might cause, but if this is not the case, I have no strong objection to using a PR as a way of doing the merge.  It would be nice if I also had the option of doing the merge from the command line, but it's not terrible if I can't.

**Comment by justinmcgrath** [comment]
Right now it isn't set up to require approvals, although that's possible. I think we always want approval on main. For develop an approval on main implies approval on develop when there are no conflicts or test failures, so for develop we could leave approvals off. We'd need to be sure to follow our guidelines without software enforcement.

However, I think it's possible give different requirements based on branch names, so maybe require approval for everything except for hotfix* branches. If that's simple enough to set up, then that would be nice to have, but I'm also happy to write down the guidelines and expect that we follow them, with the understanding that sometimes things get forgotten. I'm fairly sure you can't make a rule like "hotfixes merged to develop don't need approval when there are no conflicts or test failures, but they do need approval when there are conflicts or test failures", so the software isn't going to enforce exactly what we want anyway.

**Comment by eloch216** [comment]
> However, I think it's possible give different requirements based on branch names, so maybe require approval for everything except for hotfix* branches.

I don't think I see anything that would let us do this. I specified the following rule for `main` and `develop`: "Require a pull request before merging -- When enabled, all commits must be made to a non-protected branch and submitted via a pull request before they can be merged into a branch that matches this rule." There doesn't seem to be a way to set any exceptions to the rule, though.

I didn't set any requirements about approvals. This seems like it could be a bit annoying -- often we indicate approval with a comment but not through the official "approval" button. I think this requirement might contribute to PRs "languishing" for too long before getting merged.

**Comment by gsrohde** [comment]
> I didn't set any requirements about approvals. This seems like it could be a bit annoying -- often we indicate approval with a comment but not through the official "approval" button. I think this requirement might contribute to PRs "languishing" for too long before getting merged.

This page mentions "you can block pull requests that don't pass status checks or require that pull requests have a specific number of approving reviews before they can be merged.": https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches  But I agree that this could be annoying and am not suggesting that we implement such a requirement.

**Comment by eloch216** [comment]
> > I didn't set any requirements about approvals. This seems like it could be a bit annoying -- often we indicate approval with a comment but not through the official "approval" button. I think this requirement might contribute to PRs "languishing" for too long before getting merged.
> 
> This page mentions "you can block pull requests that don't pass status checks or require that pull requests have a specific number of approving reviews before they can be merged.": https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches But I agree that this could be annoying and am not suggesting that we implement such a requirement.

Oh, I see. Maybe we could have some type of check that would automatically pass if a hotfix branch had already been merged into `main`. That sounds like too much work to solve a problem that might not even really exist!

I think we should try using `git-flow` without too many software rules and see how it goes. We can always adjust our policies if problems start occurring. I will try writing down some guidelines and will make a PR about it soon.

**Comment by gsrohde** [comment]
> I think we should try using `git-flow` without too many software rules and see how it goes. We can always adjust our policies if problems start occurring. I will try writing down some guidelines and will make a PR about it soon.

Agreed.

**Comment by justinmcgrath** [comment]
Is everyone happy to close this after #48?

**Comment by eloch216** [comment]
> Is everyone happy to close this after #48?

I am!

**Comment by gsrohde** [comment]
> Is everyone happy to close this after #48?

Yes.
